blockstack/blockstack-browser
# Transaction Confirmation Modal

This fixes a bunch of UX issues, and potentially some bugs with the wallet. Not tracked by an issue.

### What This Does

This PR splits transactions into two actions, signing and broadcasting. Signing decrypts the user's private key, passes off the details to `blockstack.transactions.makeBitcoinSpend`, and sets the raw transaction hex to the store. The user is then prompted with a modal to confirm the transaction as decoded from the raw hex. This gives them the opportunity to see the amount sent with the fee subtracted.

To work within the limits of blockstack.js, this still retains the behavior that the amount you type in is not the amount the desired address will receive. We may want to add some language around this, either in the form or in the confirmation modal.

This also adds better loading behavior, and better error handling for amounts that are too large or too small.

### Steps to Test

* Get on an account with bitcoin in the wallet
* Attempt to send more than the balance, confirm you get an error
* Try to send an amount less than `1e-8`, e.g. `0.000000000001`, confirm you get an error
* Click the "send all" button, confirm it puts in the whole balance
* Submit the form, confirm you get a confirmation modal
* Send the transaction, confirm you get a success message
* Confirm the transaction was actually broadcast by checking your address on an explorer

### Demo

![wallet-demo](https://user-images.githubusercontent.com/649992/44274414-1a58be00-a210-11e8-8fa3-d5b0fa2b7e40.gif)
